src/main/scala/io/qbeast/spark/delta/writer/SparkDeltaDataWriter.scala

```diff
@@ -0,0 +1,194 @@
+   * Experimental: the implementation may change by using the CubeID as partition key
+   * and delegating the further execution to the underlying Format
+   *
+   * @param tableID
+   * @param schema
+   * @param data
+   * @param tableChanges
+   * @return
+   */
+  override def compact(
+      tableID: QTableID,
+      schema: StructType,
+      indexStatus: IndexStatus,
+      tableChanges: TableChanges): IISeq[FileAction] = {
+
+    val sparkSession = SparkSession.active
+
+    val cubesToCompact = indexStatus.cubesStatuses.mapValues(_.files).toIndexedSeq
+    val cubesToCompactGrouped = groupFilesToCompact(cubesToCompact)
+
+    val parallelJobCollection = new ParVector(cubesToCompactGrouped.toVector)
+
+    // Create a thread pool to be able to parallelize the compaction task
+    // Each thread will be in charge of compacting the set of files for one single cube
+    val threadPool =
+      QbeastThreadUtils.threadUtils.newForkJoinPool("Compaction", maxThreadNumber = 15)
+
+    val (factory: OutputWriterFactory, serConf: SerializableConfiguration) = {
+      val format = new ParquetFileFormat()
+      val job = Job.getInstance()
+      (
+        format.prepareWrite(sparkSession, job, Map.empty, schema),
+        new SerializableConfiguration(job.getConfiguration))
+    }
+
+    val updates =
+      try {
+        val forkJoinPoolTaskSupport = new ForkJoinTaskSupport(threadPool)
+        parallelJobCollection.tasksupport = forkJoinPoolTaskSupport
+        // For each cube with a set of files to compact, we build a different task
+        parallelJobCollection.flatMap { case (cubeId: CubeId, cubeBlocks) =>
+          if (cubeBlocks.size <= 1) { // If the number of blocks to compact is 1 or 0, we do nothing
+            Seq.empty
+          } else { // Otherwise
+            // Get the file names for the cubeId
+            val fileNames = cubeBlocks.map(f => new Path(tableID.id, f.path).toString)
+            val compactor =
+              Compactor(
+                tableID = tableID,
+                factory = factory,
+                serConf = serConf,
+                schema = schema,
+                cubeId,
+                cubeBlocks.toIndexedSeq,
+                tableChanges)
+
+            // Load the data into a DataFrame
+            // and initialize a writer in the single partition
+            sparkSession.read
+              .format("parquet")
+              .load(fileNames: _*)
+              .coalesce(1)
+              .queryExecution
+              .executedPlan
+              .execute
+              .mapPartitions(compactor.writeBlock)
+              .collect()
+              .toIndexedSeq
+          }
+
+        }.seq
+      } finally {
+        threadPool.shutdownNow()
+      }
+
+    // Return the sequence of updates
+    // which contains both Added and Removed files
+    // This code is maintained in case we need some pre-processing
+    val addedFiles = updates.collect { case a: AddFile => a }
```

Comment:
Is it possible to collect both added and removed files with a single scan?

Reply:
Actually, would be even better to return `updates` without processing anything